### ISSUE TITLE
Yield inside anonymous functions is not typed correctly

### ISSUE THREAD

**Issue by acrylic-origami** [facebook/hhvm#8405]
### HHVM Version

```
$ hhvm --version
HipHop VM 3.30.0 (rel)
Compiler: 1545075831_581101169
Repo schema: d9d101dbc225c97108a201d9b7e5a3c5c033bdda
$ hh_client --version
hh-c15d5f902880f7478651a85a635cf21d449d0b44-3.30.0 Mon Dec 17 19:46:34 2018
```

### Operating System and Version

> Ubuntu 16.04

### Standalone code, or other way to reproduce the problem

```hack
<?hh // strict
function bar(): int {
	return async { yield 42; };
}
```

### Actual result

> No errors!

### Expected result

> This is an int
> It is incompatible with an object of type AsyncGenerator

I was a bit shocked to find this out, especially with a codebase rife with one-off generators made like this. Is this known and/or feasible?

**Comment by Wilfred** [comment]
Yep, that's definitely a bug. You don't need an async block, any anonymous function seems to be sufficient:

```
function baz(): int {
  $x = () ==> { yield 'foo'; };
  return $x();
}
```

**Comment by lexidor** [comment]
`TAny` rears its ~~ugly~~ head again.
```HACK
function bar(): int {
    $x = () ==> {
        yield 'foo';
    };
    return $x();
}
```

The type of `$x` is `(function(): _)`.
`_` is never allowed to cause a typeerror.

You need to tell Hack what `$x` is to make it error.

```HACK
function bar(): int {
    $x = (): Generator<int, string, mixed> ==> {
        yield 'foo';
    };
    return $x();
}
```

```
Typing[4110] Invalid return type
   --> src/file.hack
  7 | function bar(): int {
    |                 ^^^ Expected int
  8 |     $x = (): Generator<int, string, mixed> ==> {
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ But got Generator<int, string, mixed>
 11 |     return $x();
    |            ^^^^

1 error found.
```

Inference should understand this without needing the typehint.

**Comment by Wilfred** [comment]
Fixed in 1182e4f1770f658269e706de8e2c9c762f2f9f1e.
